Allow inheritance in clades.tsv

corneliusroemer opened the pull request:
### Description of proposed changes

This PR extends `read_in_clade_definitions` in `clades.py` to allow overwritable inheritance using the following syntax:
```
    clade      gene    site     alt
    Clade_1    ctpE    81       D
    Clade_2    nuc     30642    T
    Clade_3    nuc     444296   A
    Clade_3    S       1        P
    # Clade_4 inherits from Clade_3
    Clade_4    clade   Clade_3
    Clade_4    pks8    634      T
    # Inherited allele can be overwritten
    Clade_4    S       1        L
```

Significant simplification of `clades.tsv` using this syntax can be observed in this ncov PR: https://github.com/nextstrain/ncov/pull/855/files

It uses networkx to create a directed graph of clades, to traverse in topological order and check that the graph is a DAG.

Helpful errors are raised whenever the graph is not a proper DAG.

I also enable comments in clade tsv by adding the kwarg `comment='#'` in the `pd.read_csv`

The function is rigorously unit tested by `test_clades.py` with 100% coverage, including the raising of exceptions.

Note: It may be necessary to add networkx as a dependency, if it isn't one already.

### Related issue(s)  

Resolves #823 

### Testing

Clades.py does not have any integration tests yet. I added unit tests just for the `read_in_clade_definitions` function. But given that the change is very local, I don't expect anything to break.

victorlin commented:
Tests are failing with `ModuleNotFoundError` – you'll want to add `networkx` here:

https://github.com/nextstrain/augur/blob/780e3d33edc477663d6cd2362a47b2fffcee924d/setup.py#L53-L61

corneliusroemer commented:
Thanks Victor for pointing me to where I need to change. You can do it to if you like, it's a branch on ncov itself for everyone to edit.

I'd be particularly interested in feedback on testing: is there a neater way than creating a new TSV for every test case?

Would it make sense to construct the TSVs test files programmatically to be DRYer? Or is it just fine like this, as it's nice and easy to reason?

tsibley commented:
> Would it make sense to construct the TSVs test files programmatically to be DRYer? Or is it just fine like this, as it's nice and easy to reason?

I think this is just fine and preferable over generation to me. It's easier to reason when you can look at the static file, and there's no harm in a handful of additional test data files.

victorlin commented:
> I'd be particularly interested in feedback on testing: is there a neater way than creating a new TSV for every test case?

Not following any convention, but what I did for `test_filter_groupby.py` was create valid source data (pd.DataFrame) in a [pytest fixture](https://docs.pytest.org/en/6.2.x/fixture.html):

https://github.com/nextstrain/augur/blob/780e3d33edc477663d6cd2362a47b2fffcee924d/tests/test_filter_groupby.py#L5-L15

Then "break" it in individual unit tests like so:

https://github.com/nextstrain/augur/blob/780e3d33edc477663d6cd2362a47b2fffcee924d/tests/test_filter_groupby.py#L67-L68

Though this approach isn't directly usable by your code as-is since `read_in_clade_definitions` takes a file instead of DataFrame. You could refactor to have something like `read_in_clade_definitions_from_df` as an intermediate, which is then testable using my approach, but that might be too much for this PR.

tsibley commented:
Alternatively, the TSVs are short enough that it would be reasonable IMO to inline them as strings (`"""…"""`) directly into the test file and then write them out to temp files at runtime for passing to `read_in_clade_definitions()`. (pytest offers some features for passing a ready-to-use temp file to your test functions.) This maintains the nice property @victorlin mentions of being able to see the input data right above the expected output. I would advocate against constructing a DataFrame directly, as then we're not testing the full function which in practice does actually read a TSV file.

huddlej reviewed (changes requested):
This is really cool, @corneliusroemer. The implementation is clean and the tests are easy to follow. For this type of data, I prefer the standalone TSV files to the inline data in Python because I could easily load these data into pandas from the Python prompt and test out alternate approaches to data processing.

I made a couple of suggestions below that could make `read_in_clade_definitions` a little easier to read and maintain.

corneliusroemer commented on augur/clades.py:
```diff
@@ -32,12 +45,40 @@ def read_in_clade_definitions(clade_file):
+        try:
+            G.remove_edge(root, row.clade)
+        except nx.NetworkXError:
+            raise ValueError(f"Clade {row.clade} seems to inherit from more than one clade, that's not allowed")
```
Fixed

corneliusroemer commented on augur/clades.py:
```diff
@@ -32,12 +45,40 @@ def read_in_clade_definitions(clade_file):
+        try:
+            clades[clade] = clades[next(G.predecessors(clade))].copy()
+        except StopIteration:
+            raise ValueError(f"Clade {next(G.successors(clade))} inherits from non-existent clade {clade}")
```
Implemented

corneliusroemer commented on augur/clades.py:
```diff
@@ -32,12 +45,40 @@ def read_in_clade_definitions(clade_file):
+            clades[clade][(row.gene, int(row.site)-1)] = row.alt
+    
+    # Convert items from dict of tuples to list of tuples
+    for clade in clades:
```
Implemented

corneliusroemer commented on augur/clades.py:
```diff
@@ -32,12 +45,40 @@ def read_in_clade_definitions(clade_file):
+    
+    # Convert items from dict of tuples to list of tuples
+    for clade in clades:
+        clades[clade] = list(map(lambda k: k[0] + (k[1],),clades[clade].items()))
```
Thanks, implemented

corneliusroemer commented on augur/clades.py:
```diff
@@ -32,12 +45,40 @@ def read_in_clade_definitions(clade_file):
-    for index, row in df.iterrows():
-        allele = (row.gene, row.site-1, row.alt)
-        clades[row.clade].append(allele)
-    clades.default_factory = None
```
Fixed

corneliusroemer reviewed (commented):
I think I've resolved all of your change requests @huddlej

corneliusroemer commented:
> We will need to include `networkx` as a dependency in our Bioconda recipe, too.

Yes, this is important. If we don't yet do it, it'd be good to run proper tests on our Bioconda recipe too, so we would catch things like this automatically in case we happen to forget.

corneliusroemer commented:
I've added a simple functional cram test using data from Zika. I removed all the sequences contained in the node data jsons to reduce impact on repo size.

rneher commented on augur/clades.py:
```diff
@@ -32,13 +45,51 @@ def read_in_clade_definitions(clade_file):
+    if not nx.is_directed_acyclic_graph(G):
+        raise ValueError(f"Clade definitions contain cycles {list(nx.simple_cycles(G))}")
+
+    for clade in islice(nx.topological_sort(G),1,None):
```
a few comments that explain what these functions do would be helpful. This is some sort of traversal, I guess, but can't tell from the code what kind.

corneliusroemer replied:
Do the comments I added help?

rneher commented on augur/clades.py:
```diff
@@ -32,13 +45,51 @@ def read_in_clade_definitions(clade_file):
+        raise ValueError(f"Clade definitions contain cycles {list(nx.simple_cycles(G))}")
+
+    for clade in islice(nx.topological_sort(G),1,None):
+        clades[clade] = clades[next(G.predecessors(clade))].copy()
```
same here. this is the parent?

rneher replied:
and it adds definitions from parent clade to daughter clade

rneher commented on augur/clades.py:
```diff
@@ -32,13 +45,51 @@ def read_in_clade_definitions(clade_file):
+            clades[clade][(row.gene, int(row.site)-1)] = row.alt
+    
+    # Convert items from dict[str, dict[(str,int),str]] to dict[str, list[(str,int,str)]]
+    clades = {
```
Maybe point out that mutations defined for child overwrite inherited definitions.

corneliusroemer commented:
@victorlin did you manually add networkx to the bioconda recipe or did the bioconda bump bot add the new dependency automatically?

It's definitely there which surprised me, thought I'd had to do it manually:
![image](https://user-images.githubusercontent.com/25161793/153374516-c50b2d29-3353-4e83-80e7-24720d0125de.png)

https://anaconda.org/bioconda/augur/files

tsibley commented:
@corneliusroemer Manually: https://github.com/bioconda/bioconda-recipes/commit/14321554a95aa87309ee317ae218cce1de8f45f1

victorlin commented:
Yeah, I caught the auto-bump PR bioconda/bioconda-recipes#32842 and told them to close it in favor of the manual PR.
